pkg/clusteragent/api/leader_handler.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/leaderelection"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var forwardedRequest = telemetry.NewCounterWithOpts("", "forwarded_requests",
+	[]string{"handler"}, "Counter of requests forwarded to the dca leader",
+	telemetry.Options{NoDoubleUnderscoreSep: true})
+
+// RequestPreHandler is a function that validates a http request and returns true if the request should be handled
+// otherwise it returns false and answers to the request.
+type RequestPreHandler func(w http.ResponseWriter, r *http.Request) bool
+
+// RequestHandler is a function that handles a http request.
+type RequestHandler func(w http.ResponseWriter, r *http.Request)
+
+// leaderEngine is an interface to get the leader status.
+type leaderEngine interface {
+	IsLeader() bool
+	GetLeaderIP() (string, error)
+}
+
+type leaderForwarder interface {
+	SetLeaderIP(ip string)
+	Forward(w http.ResponseWriter, r *http.Request)
+}
+
+// LeaderProxyHandler forwards queries to the leader.
+type LeaderProxyHandler struct {
+	// handlerName is used for telemetry
+	handlerName string
+	// le is the leader election engine
+	le leaderEngine
+	// leaderforwarder is used to forward queries to the leader
+	leaderForwarder leaderForwarder
+	// leaderElectionEnabled is true if leader election is enabled
+	leaderElectionEnabled bool
+	// preHandler is always called and should be used to validate the request before forwarding it or handling it
+	preHandler RequestPreHandler
+	// leaderHandler is called only by the leader
+	leaderHandler RequestHandler
+}
+
+// WithLeaderProxyHandler returns a http handler function that emits telemetry.
+func WithLeaderProxyHandler(handlerName string, preHandler RequestPreHandler, leaderHandler RequestHandler) RequestHandler {
+	leaderEngine, _ := leaderelection.GetLeaderEngine()
+	lph := LeaderProxyHandler{
+		handlerName:           handlerName,
+		leaderForwarder:       GetGlobalLeaderForwarder(),
+		le:                    leaderEngine,
+		leaderElectionEnabled: config.Datadog.GetBool("leader_election"),
+		preHandler:            preHandler,
+		leaderHandler:         leaderHandler,
+	}
+
+	return lph.handle
+}
+
+func (lph *LeaderProxyHandler) handle(w http.ResponseWriter, r *http.Request) {
+	// Perform some checks on the request to verify that it should be processed
+	if shouldBeProcessed := lph.preHandler(w, r); !shouldBeProcessed {
+		return
+	}
+	// As follower, forward to leader
+	if lph.rejectOrForwardLeaderQuery(w, r) {
+		return
+	}
+
+	// As leader, handle the request
+	lph.leaderHandler(w, r)
+}
+
+// rejectOrForwardLeaderQuery performs some checks on incoming queries that should go to a leader:
+// - Forward to leader if we are a follower
+// - Reject with "not ready" if leader election status is unknown
+func (lph *LeaderProxyHandler) rejectOrForwardLeaderQuery(rw http.ResponseWriter, req *http.Request) bool {
+	if !lph.leaderElectionEnabled {
+		return false
```

Comment:
QQ

What happens in case leader election is disabled, and language detection is enabled? 

From what I understand from the code, I assume that language detection will not work, and we will have no indicator that the reason is that leader election is disabled. 

Would it make sense to emit an ERROR here saying that leader_election should be enabled in order for language detection requests to be handled?